### PR TITLE
Implement all relevant mp functions in sp_int

### PR DESCRIPTION
**Pull request by SparkiDev**
Adds `--enable-sp-math-all` for using only SP math with support for variable key sizes and curves.

### CODE REVIEW

**Comment by guidovranken** [comment]
If you configure with `--enable-sp-math-all` but without `--enable-sp`, compilation produces errors such as:

```
wolfcrypt/src/sp_int.c:4987:18: error: array subscript is above array bounds [-Werror=array-bounds]
             t->dp[5] = l;
             ~~~~~^~~
wolfcrypt/src/sp_int.c:4992:18: error: array subscript is above array bounds [-Werror=array-bounds]
             t->dp[6] = l;
             ~~~~~^~~
wolfcrypt/src/sp_int.c:4993:18: error: array subscript is above array bounds [-Werror=array-bounds]
             t->dp[7] = h;
             ~~~~~^~~
wolfcrypt/src/sp_int.c:5081:18: error: array subscript is above array bounds [-Werror=array-bounds]
             t->dp[5] = l;
             ~~~~~^~~
wolfcrypt/src/sp_int.c:5090:18: error: array subscript is above array bounds [-Werror=array-bounds]
             t->dp[6] = l;
```

Maybe the configure script should check that `--enable-sp` is passed when `--enable-sp-math-all` is passed.

**Comment by SparkiDev** [comment]
Thanks Guido!
I've fixed this use case. Was always using --enable-sp-math-all=<bit cnt>.
Now assumes 3072 bits when RSA or DH compiled in and support for 521 bits when only ECC enabled.

**Comment by SparkiDev** [comment]
Guido has completed testing this branch and has no more bugs to report.

**Comment by toddouska** [comment]
David, please review and then assign me after you approve, thanks.

**Review by dgarske — CHANGES_REQUESTED** [review]
1) Would it be possible to use the existing --enable-sp-math option? Keep old behavior with default "yes" and use a new option like "all"?

2) Getting a cast issue due to this API difference:

```
int sp_is_bit_set(sp_int* a, unsigned int b)
int  mp_is_bit_set (mp_int * a, mp_digit b);

src/ssl.c:46208:49: error: implicit conversion loses integer precision: 'mp_digit' (aka 'unsigned long long') to 'unsigned int' [-Werror,-Wshorten-64-to-32]
    return mp_is_bit_set((mp_int*)bn->internal, (mp_digit)n);
           ~~~~~~~~~~~~~                        ^~~~~~~~~~~
```

3) DH Benchmark error -234 (yes, small or huge):

```
./configure --enable-sp --enable-sp-math-all=small --enable-all && make
$ ./wolfcrypt/benchmark/benchmark -dh
------------------------------------------------------------------------------
 wolfSSL version 4.5.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
DH      2048 key gen         0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH key gen 2048 failed: -234
DH      2048 agree           0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH agree 2048 failed: -234
```

4) Scan-build warnings `scan-build ./configure --enable-sp --enable-sp-math-all=small && scan-build make`:

```
wolfcrypt/src/sp_int.c:2235:18: warning: The left expression of the compound assignment is an uninitialized value. The computed value will also be garbage
        r->dp[0] &= (sp_int_digit)0 - (a->used != 0);
        ~~~~~~~~ ^
wolfcrypt/src/sp_int.c:3301:13: warning: Value stored to 'tt' is never read
            tt += tr / d;
            ^     ~~~~~~
wolfcrypt/src/sp_int.c:12098:9: warning: Value stored to 'l' is never read
        l = 0;
        ^   ~
3 warnings generated.
```

5) smallstack build error:

```
./configure --enable-sp --enable-sp-math-all=small --enable-smallstack && make
wolfcrypt/src/sp_int.c:2329:15: error: use of undeclared identifier 'FP_MEM'
       return FP_MEM;
              ^
1 error generated.
```

6) Note I am not seeing much performance difference between `--enable-sp-math-all=yes` and --enable-sp-math-all=huge`? Is that expected?

```
--enable-sp-math-all=yes
RSA     1024 key gen        59 ops took 1.030 sec, avg 17.451 ms, 57.304 ops/sec
RSA     2048 key gen         6 ops took 1.089 sec, avg 181.520 ms, 5.509 ops/sec
RSA     2048 public      21500 ops took 1.004 sec, avg 0.047 ms, 21417.049 ops/sec
RSA     2048 private       500 ops took 1.021 sec, avg 2.043 ms, 489.489 ops/sec
DH      2048 key gen         0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH key gen 2048 failed: -234
DH      2048 agree           0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH agree 2048 failed: -234
ECC      256 key gen      2858 ops took 1.000 sec, avg 0.350 ms, 2857.243 ops/sec
ECDHE    256 agree       10200 ops took 1.008 sec, avg 0.099 ms, 10123.718 ops/sec
ECDSA    256 sign         8300 ops took 1.011 sec, avg 0.122 ms, 8212.700 ops/sec
ECDSA    256 verify       9200 ops took 1.007 sec, avg 0.109 ms, 9137.236 ops/sec
ECC      256 encryption    9900 ops took 1.007 sec, avg 0.102 ms, 9827.394 ops/sec
ECC      256 decryption    9900 ops took 1.008 sec, avg 0.102 ms, 9817.893 ops/sec
CURVE  25519 key gen     19578 ops took 1.000 sec, avg 0.051 ms, 19577.981 ops/sec
CURVE  25519 agree       19600 ops took 1.001 sec, avg 0.051 ms, 19574.318 ops/sec
ED     25519 key gen     44776 ops took 1.000 sec, avg 0.022 ms, 44775.509 ops/sec
ED     25519 sign        39800 ops took 1.002 sec, avg 0.025 ms, 39700.784 ops/sec
ED     25519 verify      14500 ops took 1.007 sec, avg 0.069 ms, 14405.714 ops/sec
CURVE    448 key gen      3676 ops took 1.000 sec, avg 0.272 ms, 3675.107 ops/sec
CURVE    448 agree        3700 ops took 1.011 sec, avg 0.273 ms, 3660.543 ops/sec
ED       448 key gen      9530 ops took 1.000 sec, avg 0.105 ms, 9529.314 ops/sec
ED       448 sign         9300 ops took 1.003 sec, avg 0.108 ms, 9269.548 ops/sec
ED       448 verify       3000 ops took 1.017 sec, avg 0.339 ms, 2948.534 ops/sec
```

```
--enable-sp-math-all=huge
RSA     1024 key gen        55 ops took 1.001 sec, avg 18.194 ms, 54.964 ops/sec
RSA     2048 key gen        10 ops took 1.168 sec, avg 116.824 ms, 8.560 ops/sec
RSA     2048 public      21700 ops took 1.002 sec, avg 0.046 ms, 21650.637 ops/sec
RSA     2048 private       500 ops took 1.015 sec, avg 2.030 ms, 492.619 ops/sec
DH      2048 key gen         0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH key gen 2048 failed: -234
DH      2048 agree           0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH agree 2048 failed: -234
ECC      256 key gen      2871 ops took 1.000 sec, avg 0.348 ms, 2870.750 ops/sec
ECDHE    256 agree        9600 ops took 1.002 sec, avg 0.104 ms, 9577.043 ops/sec
ECDSA    256 sign         7900 ops took 1.007 sec, avg 0.128 ms, 7843.114 ops/sec
ECDSA    256 verify       8900 ops took 1.000 sec, avg 0.112 ms, 8898.540 ops/sec
ECC      256 encryption    9400 ops took 1.007 sec, avg 0.107 ms, 9334.972 ops/sec
ECC      256 decryption    9600 ops took 1.006 sec, avg 0.105 ms, 9541.644 ops/sec
CURVE  25519 key gen     18778 ops took 1.000 sec, avg 0.053 ms, 18777.830 ops/sec
CURVE  25519 agree       18400 ops took 1.002 sec, avg 0.054 ms, 18359.113 ops/sec
ED     25519 key gen     43564 ops took 1.000 sec, avg 0.023 ms, 43563.138 ops/sec
ED     25519 sign        39600 ops took 1.001 sec, avg 0.025 ms, 39562.732 ops/sec
ED     25519 verify      14800 ops took 1.006 sec, avg 0.068 ms, 14710.781 ops/sec
CURVE    448 key gen      3806 ops took 1.000 sec, avg 0.263 ms, 3805.870 ops/sec
CURVE    448 agree        3700 ops took 1.017 sec, avg 0.275 ms, 3639.314 ops/sec
ED       448 key gen      9183 ops took 1.000 sec, avg 0.109 ms, 9182.477 ops/sec
ED       448 sign         9000 ops took 1.000 sec, avg 0.111 ms, 8995.987 ops/sec
ED       448 verify       3000 ops took 1.001 sec, avg 0.334 ms, 2996.204 ops/sec
```

```
--enable-sp-math-all=small
RSA     1024 key gen        58 ops took 1.011 sec, avg 17.438 ms, 57.345 ops/sec
RSA     2048 key gen         3 ops took 1.154 sec, avg 384.681 ms, 2.600 ops/sec
RSA     2048 public      10400 ops took 1.008 sec, avg 0.097 ms, 10320.359 ops/sec
RSA     2048 private       200 ops took 1.149 sec, avg 5.746 ms, 174.039 ops/sec
DH      2048 key gen         0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH key gen 2048 failed: -234
DH      2048 agree           0 ops took 0.000 sec, avg 0.000 ms, 0.000 ops/sec
Benchmark DH agree 2048 failed: -234
ECC      256 key gen       529 ops took 1.001 sec, avg 1.893 ms, 528.288 ops/sec
ECDHE    256 agree        1300 ops took 1.011 sec, avg 0.778 ms, 1285.268 ops/sec
ECDSA    256 sign         1200 ops took 1.039 sec, avg 0.866 ms, 1154.443 ops/sec
ECDSA    256 verify        700 ops took 1.107 sec, avg 1.582 ms, 632.203 ops/sec
ECC      256 encryption    1300 ops took 1.035 sec, avg 0.796 ms, 1256.466 ops/sec
ECC      256 decryption    1100 ops took 1.050 sec, avg 0.954 ms, 1047.871 ops/sec
CURVE  25519 key gen     17192 ops took 1.000 sec, avg 0.058 ms, 17192.000 ops/sec
CURVE  25519 agree       17500 ops took 1.005 sec, avg 0.057 ms, 17410.340 ops/sec
ED     25519 key gen     41366 ops took 1.000 sec, avg 0.024 ms, 41365.329 ops/sec
ED     25519 sign        32200 ops took 1.002 sec, avg 0.031 ms, 32131.467 ops/sec
ED     25519 verify      10900 ops took 1.004 sec, avg 0.092 ms, 10851.549 ops/sec
CURVE    448 key gen      3092 ops took 1.000 sec, avg 0.323 ms, 3091.301 ops/sec
CURVE    448 agree        3600 ops took 1.021 sec, avg 0.283 ms, 3527.437 ops/sec
ED       448 key gen      8502 ops took 1.000 sec, avg 0.118 ms, 8501.159 ops/sec
ED       448 sign         7400 ops took 1.006 sec, avg 0.136 ms, 7358.821 ops/sec
ED       448 verify       2600 ops took 1.050 sec, avg 0.404 ms, 2476.846 ops/sec
```

Thanks, David

File: wolfssl/wolfcrypt/sp_int.h [hunk 1]
```diff
@@ -128,184 +208,581 @@ typedef struct sp_ecc_ctx {
...
+#define sp_print(a, s)                                  \
+    do {                                                \
+        int ii;                                         \
+        fprintf(stderr, "%s=0x", s);                    \
+        if ((a)->sign == MP_NEG) {                      \
+            fprintf(stderr, "-");                       \
+        }                                               \
+        fprintf(stderr, "0");                           \
+        for (ii = (a)->used-1; ii >= 0; ii--) {         \
+            fprintf(stderr, SP_PRINT_FMT, (a)->dp[ii]); \
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/sp_int.h:320]: Change to detect size of unsigned long and use appropriate format string.

File: wolfssl/wolfcrypt/sp_int.h [hunk 2]
```diff
@@ -128,184 +208,581 @@ typedef struct sp_ecc_ctx {
...
+
+/* Print out a single multi-precision digit.
+ *
+ * @param  [in]  a  SP integer digit to print.
+ * @param  [in]  s  String that describes the use of the number.
+ */
+#define sp_print_digit(a, s)                            \
+    do {                                                \
+        fprintf(stderr, "%s=0x0", s);                   \
+        fprintf(stderr, SP_PRINT_FMT, a);               \
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/sp_int.h:335]: Change to detect size of unsigned long and use appropriate format string.

File: tests/api.c [hunk 3]
```diff
@@ -14339,12 +14339,12 @@ static int test_wc_RsaKeyToPublicDer (void)
...
     WC_RNG      rng;
     byte*       der;
-    #ifndef WOLFSSL_SP_MATH
+    #if !defined(WOLFSSL_SP_MATH) || defined(WOLFSSL_SP_MATH_ALL)
     int         bits = 1024;
     word32      derLen = 162;
     #else
     int         bits = 2048;
-    word32      derLen = 290;
+    word32      derLen = 294;
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply tests/api.c:14347]: Not sure why I did this?!?
Returned it to 290.

File: wolfcrypt/src/tfm.c [hunk 4]
```diff
@@ -3588,6 +3589,38 @@ int fp_read_unsigned_bin(fp_int *a, const unsigned char *b, int c)
...
+     unsigned char *pd = (unsigned char *)a->dp;
+
+     a->used = (c + sizeof(fp_digit) - 1)/sizeof(fp_digit);
+     /* read the bytes in */
+#ifdef BIG_ENDIAN_ORDER
+     {
+       /* Use Duff's device to unroll the loop. */
+       int idx = (c - 1) & ~7;
+       switch (c % 8) {
+       case 0:    do { pd[idx+0] = *b++; // fallthrough
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/tfm.c:3605]: changed

**Comment by SparkiDev** [comment]
configure --enable-sp --enable-sp-math-all=huge will use the sp_c64.c file for 2048 bit RSA and DH.

**Comment by SparkiDev** [comment]
John changed GeneratePrivateDh() to check incoming privSz.
benchmark.c sets privSz to 0.
Changed benchmark.c in this pull request to set privSz to size of buffer.

**Comment by SparkiDev** [comment]
mp_is_bit_set macro changed to cast from mp_digit to unsigned int.

**Comment by SparkiDev** [comment]
./configure --enable-sp --enable-sp-math-all=small --enable-smallstack && make
now builds and passes tests

**Comment by SparkiDev** [comment]
scan-build issues fixed.
Either bogus or unused initialization that makes no difference.

**Comment by SparkiDev** [comment]
I expect the sp-math-all to change to something else when it is the default.
Having sp in the name is confusing but needed for now to distinguish implementations.

**Review by dgarske — CHANGES_REQUESTED** [review]
Hi Sean,

Still getting several errors on the sp_print. If I use `%016llx` it builds fine on my Mac with apple clang. 

```
wolfcrypt/src/sp_int.c:3422:9: error: format specifies type 'unsigned long' but the argument has type 'sp_int_digit' (aka 'unsigned long long') [-Werror,-Wformat]
        sp_print(a, "a");
        ^~~~~~~~~~~~~~~~
./wolfssl/wolfcrypt/sp_int.h:309:43: note: expanded from macro 'sp_print'
            fprintf(stderr, SP_PRINT_FMT, (a)->dp[ii]); \
                            ~~~~~~~~~~~~  ^~~~~~~~~~~
```

Otherwise all looks and tests out correctly now. Thanks!

**Comment by SparkiDev** [comment]
Print format fixed and reliance on stdint.h removed.

**Review by dgarske — APPROVED** [review]
I tested on several embedded targets and fixed a few minor issues. This PR is approved. Over to Todd for final.

**Comment by toddouska** [comment]
Looks good @SparkiDev !  Please resolve merge conflicts and I'll approve / merge my morning.  Thanks.

**Review by toddouska — APPROVED** [review]

**Pull request merged**